src/analyses/variable-sensitivity/value_set_abstract_object.cpp

```diff
@@ -0,0 +1,305 @@
+#include <analyses/variable-sensitivity/constant_abstract_value.h>
+#include <analyses/variable-sensitivity/constant_pointer_abstract_object.h>
+#include <analyses/variable-sensitivity/context_abstract_object.h>
+#include <analyses/variable-sensitivity/interval_abstract_value.h>
+#include <analyses/variable-sensitivity/struct_abstract_object.h>
+#include <analyses/variable-sensitivity/union_abstract_object.h>
+#include <analyses/variable-sensitivity/value_set_abstract_object.h>
+
+value_set_abstract_objectt::value_set_abstract_objectt(const typet &type)
+  : abstract_valuet(type), my_type(type_to_abstract_type(type))
+{
+  switch(my_type)
+  {
+  case abstract_typet::POINTER:
+    values.insert(std::make_shared<constant_pointer_abstract_objectt>(type));
+    break;
+  case abstract_typet::CONSTANT:
+    values.insert(std::make_shared<constant_abstract_valuet>(type));
+    break;
+  case abstract_typet::UNSUPPORTED:
+    UNREACHABLE;
+  }
+  verify();
+}
+
+value_set_abstract_objectt::value_set_abstract_objectt(
+  const typet &type,
+  bool top,
+  bool bottom)
+  : abstract_valuet(type, top, bottom), my_type(type_to_abstract_type(type))
+{
+  switch(my_type)
+  {
+  case abstract_typet::POINTER:
+    values.insert(
+      std::make_shared<constant_pointer_abstract_objectt>(type, top, bottom));
+    break;
+  case abstract_typet::CONSTANT:
+    values.insert(
+      std::make_shared<constant_abstract_valuet>(type, top, bottom));
+    break;
+  case abstract_typet::UNSUPPORTED:
+    UNREACHABLE;
+  }
+  verify();
+}
+
+value_set_abstract_objectt::value_set_abstract_objectt(
+  const exprt &expr,
+  const abstract_environmentt &environment,
+  const namespacet &ns)
+  : abstract_valuet(expr.type(), false, false),
+    my_type(type_to_abstract_type(expr.type()))
+{
+  switch(my_type)
+  {
+  case abstract_typet::POINTER:
+    values.insert(std::make_shared<constant_pointer_abstract_objectt>(
+      expr, environment, ns));
+    break;
+  case abstract_typet::CONSTANT:
+    values.insert(
+      std::make_shared<constant_abstract_valuet>(expr, environment, ns));
+    break;
+  case abstract_typet::UNSUPPORTED:
+    UNREACHABLE;
+  }
+  verify();
+}
+
+abstract_object_pointert value_set_abstract_objectt::expression_transform(
+  const exprt &expr,
+  const std::vector<abstract_object_pointert> &operands,
+  const abstract_environmentt &environment,
+  const namespacet &ns) const
+{
+  std::size_t num_operands = expr.operands().size();
+  PRECONDITION(operands.size() == num_operands);
+
+  std::vector<abstract_object_sett> collective_operands;
+  collective_operands.reserve(num_operands);
+  for(const auto &op : operands)
+  {
+    auto vsab = std::dynamic_pointer_cast<const value_set_abstract_objectt>(
+      maybe_unwrap_context(op));
+    INVARIANT(vsab, "should be a value set abstract object");
+    collective_operands.push_back(vsab->get_values());
+  }
+
+  abstract_object_sett resulting_objects;
+  for_each_comb(
```

Comment:
This bit needs to account for the possibility of an operand being top (or bottom I guess, although in that case I don't think we'd even evaluate this?).

Reply:
It makes sense to me that that this should cope with top, but I have no idea how to implement it